Keep open page cursors

lutovich opened the pull request:
Previously most of Kernel API cursors directly used stores to read records. This means they were opening page cursors all the time and pinning/unpinning page cache pages.

This PR makes all Kernel API cursors use record cursors. Record cursors keep underlying page cache cursors open as long as possible thus avoiding excessive pins/unpins. Record cursors are also kept open and pooled with KTIs as part of their StoreStatements.

tinwelint committed:
Reuse record cursors in StoreStatement

Previously most of Kernel API cursors directly used stores to read
records. This means they were opening page cursors all the time and
pinning/unpinning page cache pages.

This commit makes all Kernel API cursors use record cursors. Record
cursors keep underlying page cache cursors open as long as possible thus
avoiding excessive pins/unpins. Record cursors are also kept open and
pooled with KTIs as part of their StoreStatements.

lutovich committed:
Small cleanups around cursors, couple tests added

Co-authored-by: @burqen

lutovich commented:
@tinwelint `StoreAbstractNodeCursor#labels()` and `StoreAbstractNodeCursor#label(int)` do not use cursors. they eagerly read all dynamic label records (if any) into an array and then there is basically a cursor view over that array. do you think we should make these places use record cursors as well?

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreStatement.java:
```diff
@@ -59,63 +57,66 @@
     private final Supplier<LabelScanReader> labelScanStore;
     private LabelScanReader labelScanReader;
-    private boolean closed;
+    private boolean acquired, closed;
```
Please split this into two distinct field definitions.

chrisvest commented on community/kernel/src/main/java/org/neo4j/storageengine/api/StorageStatement.java:
```diff
@@ -33,20 +33,28 @@
+ * Creating and closing {@link StorageStatement} and there's also benefits keeping these statements opened
+ * during a longer period of time, with the assumption that it's still one thread at a time using each.
+ * With that in mind these statements should not be opened and closed for each operation, perhaps not even
+ * for each transaction.
```
> With that in mind these statements should not be opened and closed for each operation, perhaps not even
> for each transaction.

How does that interact with transaction state?

lutovich replied:
I think `StorageStatement` is only responsible for providing cursors over data in the store. Those cursors are later augmented with similar cursors from the transaction state (in the cake). Do you think this should be added to the javadoc?

chrisvest replied:
Yes, a quick reminder that we are below the transaction state in the stack would be good, I think.

lutovich commented:
Marked as not ready for merge because we also need to use record cursor for dynamic labels.

lutovich committed:
Read dynamic labels through record cursor

Record cursor for dynamic labels is now pooled together with other cursors.
This is done to reuse same page cursor when reading labels of nodes inside the
same transaction.

lutovich committed:
Use property and label cursors simultaneously

This is basically a revert of 28857cf43af643a7ec9b044bc128e02d5267cf1e.
Now it is possible to use node property and node label cursors at the
same time to update indexes.

lutovich commented:
@chrisvest fixed javadoc and added two more commits.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreStatement.java:
```diff
@@ -159,18 +162,33 @@ public void acquire()
         if ( indexReaderFactory != null )
         {
             indexReaderFactory.close();
+            // we can actually keep this object around
```
Why?

lutovich replied:
Implementation would be `IndexReaderFactory.Caching`. It's `#close()` closes all cached index readers and clears map used to keep store them.

chrisvest replied:
And we never need to reinitialise it?

lutovich replied:
Yes, just clear the internal map and that is it

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/RecordCursor.java:
```diff
@@ -65,6 +77,31 @@
+    boolean next( long id, R record, RecordLoad mode );
+
+    @SuppressWarnings( "unchecked" )
+    default List<R> getAll()
```
`getAll` has no javadoc

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/RecordCursors.java:
```diff
@@ -0,0 +1,105 @@
+
+import static org.neo4j.kernel.impl.store.record.RecordLoad.NORMAL;
+
+public class RecordCursors implements AutoCloseable
```
Can we give this guy a bit of javadoc?

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java:
```diff
@@ -42,17 +42,21 @@
+    public StorePropertyCursor(
+            RecordCursor<PropertyRecord> recordCursor,
+            RecordCursor<DynamicRecord> stringRecordCursor,
+            RecordCursor<DynamicRecord> arrayRecordCursor,
```
This guy could alternatively take in the `RecordCursors` object, and then grab the cursors it needs in the constructor.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyPayloadCursor.java:
```diff
@@ -286,25 +273,23 @@ private void readFromStore( RecordCursor<DynamicRecord> cursor )
+        while ( true )
         {
-            while ( true )
+            cursor.next();
```
The result of `cursor.next()` is not checked.

lutovich replied:
This is on purpose here because we need to read notInUse records.

chrisvest replied:
Ah, okay, yes I see it's given a FORCE load mode.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreAbstractNodeCursor.java:
```diff
@@ -234,7 +239,7 @@ public boolean next()
                     }
 
-                    relationshipGroupStore.getRecord( groupId, group, NORMAL );
+                    cursors.relationshipGroup().next( groupId, group, NORMAL );
```
Result of `next()` not checked.

lutovich replied:
`RecordCursor` can return `false` only when read record is not in use. After reading, each record is verified by the `LoadMode`. `LoadMode.NORMAL` throws `InvalidRecordException` if record is not in use. So here and in all other places where we read with `NORMAL` check of `next()` result could be skipped because mode would throw if unused record was read.

chrisvest replied:
The `RecordCursor` API is quite confusing in this regard.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreAbstractNodeCursor.java:
```diff
@@ -304,10 +309,11 @@ public int degree( Direction direction )
             {
-                relationshipGroupStore.getRecord( groupId, group, NORMAL );
-                count += nodeDegreeByDirection( group, direction );
+                cursors.relationshipGroup().next( groupId, group, NORMAL );
```
Result of `next()` not checked.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreAbstractNodeCursor.java:
```diff
@@ -333,12 +339,13 @@ public int degree( Direction direction, int relType )
             while ( groupId != Record.NO_NEXT_RELATIONSHIP.intValue() )
             {
-                relationshipGroupStore.getRecord( groupId, group, NORMAL );
+                cursors.relationshipGroup().next( groupId, group, NORMAL );
```
Result of `next()` not checked.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreAbstractNodeCursor.java:
```diff
@@ -397,31 +404,31 @@ public boolean isDense()
         }
-        RelationshipRecord record = relationshipStore.getRecord( relationshipId,
-                relationshipStore.newRecord(), NORMAL );
+        cursors.relationship().next( relationshipId, record, NORMAL );
```
Result of `next()` not checked.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreAbstractNodeCursor.java:
```diff
@@ -533,11 +541,11 @@ public boolean next()
             if ( groupId != Record.NO_NEXT_RELATIONSHIP.intValue() )
             {
-                relationshipGroupStore.getRecord( groupId, group, NORMAL );
+                cursors.relationshipGroup().next( groupId, group, NORMAL );
```
Result of `next()` not checked.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreNodeRelationshipCursor.java:
```diff
@@ -85,7 +86,7 @@ public StoreNodeRelationshipCursor init( boolean isDense,
         if ( isDense && relationshipId != Record.NO_NEXT_RELATIONSHIP.intValue() )
         {
-            relationshipGroupStore.getRecord( firstRelId, groupRecord, FORCE );
+            cursors.relationshipGroup().next( firstRelId, groupRecord, FORCE );
```
Result of `next()` not checked.

lutovich replied:
`next()` returns `record.inUse()` but here and in other places we read with `FORCE` mode so we need to read notInUse records and do not really care about `next()` return value

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreNodeRelationshipCursor.java:
```diff
@@ -101,7 +102,7 @@ public boolean next()
         while ( relationshipId != NO_NEXT_RELATIONSHIP.intValue() )
         {
-            relationshipStore.getRecord( relationshipId, relationshipRecord, FORCE );
+            relationshipRecordCursor.next( relationshipId, relationshipRecord, FORCE );
```
Result of `next()` not checked.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreNodeRelationshipCursor.java:
```diff
@@ -207,7 +208,7 @@ private long nextChainStart()
                 if ( !NULL_REFERENCE.is( groupRecord.getNext() ) )
                 {
-                    relationshipGroupStore.getRecord( groupRecord.getNext(), groupRecord, FORCE );
+                    cursors.relationshipGroup().next( groupRecord.getNext(), groupRecord, FORCE );
```
Result of `next()` not checked.

lutovich committed:
Couple fixes after the code review

Simplified StorePropertyCursor construction, added javadocs.

lutovich commented:
@chrisvest i think most of your comments are fixed. i did not change places where result of `next()` is not checked because they seem to do exactly what we want. combination of `LoadMode` and `RecordCursor` is quite convoluted. maybe we can revisit it as a separate task, wdyt?
